### PR TITLE
A central ServletContextListener to implement dependency injection pattern

### PR DESCRIPTION
**Pull request by lzoebaker**
The ServletContextListener calls constructors of all the other servlets with Datastore/Userservice as an argument. From research, this is the way to implement dependency injection with servlets as servlet constructors will not be invoked by themselves.

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/AuthenticationServlet.java [hunk 1]
```diff
@@ -13,15 +13,20 @@
...
-  private final UserService userService = UserServiceFactory.getUserService(); 
-  private final UserDatabase userDatabase = new UserDatabase(datastore);
+  private final DatastoreService datastore;
+  private final UserService userService;
+  private final UserDatabase userDatabase;
+
+  public AuthenticationServlet(DatastoreService datastore, UserService userService) {
+     this.datastore = datastore;
+     this.userService = userService;
+     this.userDatabase = new UserDatabase(this.datastore);
```
Comment (WillHipschman) [inline portfolio/src/main/java/com/google/sps/servlets/AuthenticationServlet.java:28]: Should you inject user database too?

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 2]
```diff
@@ -40,14 +37,17 @@
   private static final String MAX_FORM_SUBMIT = "max-comment-submit";
   private static final int MAX_COMMENT_DEFAULT = 12;
   private int maxCommentsToDisplay;
-  private final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-  private final UserService userService = UserServiceFactory.getUserService(); 
-  private final CommentDatabase commentDatabase = new CommentDatabase(this.datastore);
-  private final UserDatabase userDatabase = new UserDatabase(this.datastore);
+  private CommentDatabase commentDatabase;
```
Comment (WillHipschman) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:40]: These should be final

File: portfolio/src/main/java/com/google/sps/servlets/DisplayNameServlet.java [hunk 3]
```diff
@@ -14,13 +14,18 @@
...
-  private final UserDatabase userDatabase = new UserDatabase(this.datastore);
-  private final UserService userService = UserServiceFactory.getUserService();
+  private final DatastoreService datastore;
+  private final UserDatabase userDatabase;
+  private final UserService userService;
+
+  public DisplayNameServlet(DatastoreService datastore, UserService userService) {
+     this.datastore = datastore;
+     this.userService = userService;
+     this.userDatabase = new UserDatabase(this.datastore);
```
Comment (WillHipschman) [inline portfolio/src/main/java/com/google/sps/servlets/DisplayNameServlet.java:27]: Inject UserDatabase

File: portfolio/src/main/java/com/google/sps/servlets/Application.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+          .addMapping("/delete-comment");
+      context.addServlet("displayNameServlet", new DisplayNameServlet(datastore, userService, userDatabase))
+          .addMapping("/display-name");
+      context.addServlet("authenticationServlet", new AuthenticationServlet(datastore, userService, userDatabase))
+          .addMapping("/login");
+    }
+    @Override
+    public void contextDestroyed(ServletContextEvent servletContextEvent) {
+    }
+};
```
Comment (menynexi) [inline portfolio/src/main/java/com/google/sps/servlets/Application.java:36]: new line should be added

**Pull request closed without merging**